src/rust/engine/process_execution/src/nailgun/mod.rs

```diff
@@ -0,0 +1,254 @@
+            python_distribution_absolute_path: python_distribution_absolute_path,
+        }
+    }
+
+    // Ensure that the workdir for the given nailgun name exists.
+    fn get_nailguns_workdir(&self, nailgun_name: &NailgunProcessName) -> Result<PathBuf, String> {
+        let workdir = self.workdir_base.clone().join(nailgun_name);
+        if self.workdir_base.exists() {
+            debug!("Creating nailgun workdir at {:?}", self.workdir_base);
+            std::fs::create_dir_all(workdir.clone())
+                .map_err(|err| format!("Error creating the nailgun workdir! {}", err))
+                .map(|_| workdir)
+        } else {
+            debug!("nailgun workdir {:?} exits. Reusing that...", self.workdir_base);
+            Ok(workdir)
+        }
+    }
+
+    fn calculate_nailgun_name(main_class: &String, digest: &Digest) -> String {
+        format!("{}_{}", main_class, digest.0)
+    }
+}
+
+impl super::CommandRunner for NailgunCommandRunner {
+    fn run(
+        &self,
+        req: MultiPlatformExecuteProcessRequest,
+        context: Context) -> BoxFuture<FallibleExecuteProcessResult, String> {
+
+        let mut client_req = self.extract_compatible_request(&req).unwrap();
+
+        if !client_req.is_nailgunnable {
+            trace!("The request is not nailgunnable! Short-circuiting to regular process execution");
+            return self.inner.run(req, context)
+        }
+        debug!("Running request under nailgun:\n {:#?}", &client_req);
+
+        // Separate argument lists, to form distinct EPRs for (1) starting the nailgun server and (2) running the client in it.
+        let ParsedArgLists {nailgun_args, client_args } = split_args(&client_req.argv);
+        let nailgun_req = get_nailgun_request(nailgun_args, client_req.input_files, client_req.jdk_home.clone());
+        trace!("Extracted nailgun request:\n {:#?}", &nailgun_req);
+
+        let maybe_jdk_home = nailgun_req.jdk_home.clone();
+
+        let main_class = client_args.iter().next().unwrap().clone(); // We assume the last one is the main class name
+        let nailgun_req_digest = crate::digest(MultiPlatformExecuteProcessRequest::from(nailgun_req.clone()), &self.metadata);
+        let nailgun_name = NailgunCommandRunner::calculate_nailgun_name(&main_class, &nailgun_req_digest);
+        let nailgun_name2 = nailgun_name.clone();
+        let nailgun_name3 = nailgun_name.clone();
+
+        let nailguns_workdir = try_future!(self.get_nailguns_workdir(&nailgun_name));
+
+        // Materialize the directory for running the nailgun server, if we need to.
+        let workdir_path2 = nailguns_workdir.clone();
+        let workdir_path3= nailguns_workdir.clone();
+        let workdir_path4 = nailguns_workdir.clone();
+        let materialize = self.inner
+            .store
+            // TODO This materializes the input files in the client req, which is a superset of the files we need (we only need the classpath, not the input files)
```

Comment:
For now, can you set the `input_files` on the server request to be the same as the client request (so that this code will correctly use `nailgun_req.input_files`), and we can just address the TODO by fixing how we populate the `input_files` on the `nailgun_req`, rather than also needing to change this code?

Reply:
I cannot, because that input_files will include the source files. Since we calculate the fingerprint by hashing the request (I think this a solid way of doing it, but open to suggestions), the hash will be different every time because the source files will be different every time. Opened issue and linked it though.